edx_rest_framework_extensions/auth/jwt/authentication.py

```diff
@@ -216,6 +229,78 @@ def is_authenticating_with_jwt_cookie(cls, request):
         except Exception:  # pylint: disable=broad-exception-caught
             return False
 
+    def _is_jwt_cookie_and_session_user_mismatched(self, request, jwt_user=None):
+        """
+        Returns True if JWT cookie and session user do not match, False otherwise.
+
+        Arguments:
+            request: The request.
+            jwt_user (User): The valid JWT user. If not user is supplied, it is assumed that
+                the cookie was invalid, and we attempt to get the user_id from the invalid
+                token.
+
+        Other notes:
+        - If ENABLE_JWT_VS_SESSION_USER_CHECK is toggled off, always return False.
+        - Also adds monitoring details for mismatches.
+        - Should only be called for JWT cookies.
+        """
+        is_jwt_vs_session_user_check_enabled = get_setting(ENABLE_JWT_VS_SESSION_USER_CHECK)
+        # .. custom_attribute_name: is_jwt_vs_session_user_check_enabled
+        # .. custom_attribute_description: This is temporary custom attribute to show
+        #      whether ENABLE_JWT_VS_SESSION_USER_CHECK is toggled on or off.
+        set_custom_attribute('is_jwt_vs_session_user_check_enabled', is_jwt_vs_session_user_check_enabled)
+        if not is_jwt_vs_session_user_check_enabled:
+            return False
+
+        has_request_user = (
+            hasattr(request, '_request') and hasattr(request._request, 'user')  # pylint: disable=protected-access
+        )
+        if not has_request_user:  # pragma: no cover
+            # .. custom_attribute_name: jwt_auth_request_user_not_found
+            # .. custom_attribute_description: This custom attribute will show that we
+            #      were unable to find the session user. This should not occur outside
+            #      of tests, because there should still be an unauthenticated user, but
+            #      this attribute could be used to check for the unexpected.
+            set_custom_attribute('jwt_auth_request_user_not_found', True)
+            return False
+
+        wsgi_request_user = request._request.user  # pylint: disable=protected-access
+        if wsgi_request_user and wsgi_request_user.is_authenticated:
+            session_user_id = wsgi_request_user.id
+        else:
+            session_user_id = None
+
+        if jwt_user:
+            jwt_user_id = jwt_user.id
+        else:
+            cookie_token = JSONWebTokenAuthentication.get_token_from_cookies(request.COOKIES)
+            invalid_decoded_jwt = unsafe_jwt_decode_handler(cookie_token)
+            jwt_user_id = invalid_decoded_jwt['user_id']
```

Comment:
I originally though exceptions wouldn't be so bad, because the auth has failed anyway. However, I agree that some of this is more about monitoring, so I made this all more resilient. I also refactored a bit to make the code a bit more clear.